File: src/contest/ioi/IOI_2015_Sorting.cpp
#include <bits/stdc++.h>

#define mp make_pair
#define pb push_back
#define INF 1<<30
#define MOD 1000000007
#define rint(x) scanf("%d", &(x))
#define rlong(x) scanf("%lld", &(x))
#define SIZE 200000
#define l(x) x << 1
#define r(x) x << 1|1
#define m(x, y) (x + y)/2

using namespace std;

typedef long long ll;
typedef pair<int, int> pi;
typedef pair<ll, ll> pll;

int val[SIZE], valIndex[SIZE];
int sorted[SIZE];
unordered_map<int, int> toIndex;

int findSwapPairs (int N, int S[], int M, int X[], int Y[], int P[], int Q[]) {
    for (int i = 0; i < N; i++) {
        val[i] = S[i];
        sorted[i] = val[i];
        valIndex[i] = i;
        toIndex[val[i]] = i;
    }

    sort(sorted, sorted + N);

    int lo = 0, hi = N - 1;
    vector<int> s1, s2;

    while (lo <= hi) {
        int mid = (hi + lo) / 2;

        for (int i = mid - 1; i >= 0; i--)
            swap(valIndex[X[i]], valIndex[Y[i]]);

        int cnt = 0;
        for (int i = 0; i < N; i++) {
            if (sorted[valIndex[i]] != val[i]) {
                int j = toIndex[sorted[valIndex[i]]];

                s1.pb(i);
                s2.pb(j);
                swap(val[i], val[j]);
                toIndex[val[j]] = j;
                toIndex[val[i]] = i;
                cnt++;
            }
        }

        for (int i = s1.size() - 1; i >= 0; i--) {
            swap(val[s1[i]], val[s2[i]]);
            toIndex[val[s1[i]]] = s1[i];
            toIndex[val[s2[i]]] = s2[i];
        }
        for (int i = 0; i < mid; i++)
            swap(valIndex[X[i]], valIndex[Y[i]]);
        if (cnt <= mid)
            hi = mid - 1;
        else
            lo = mid + 1;
        s1.clear();
        s2.clear();
    }

    // rebuilding sequence

    vector<int> ans1, ans2;

    for (int i = lo - 1; i >= 0; i--)
        swap(valIndex[X[i]], valIndex[Y[i]]);

    for (int i = 0; i < N; i++) {
        if (sorted[valIndex[i]] != val[i]) {
            int j = toIndex[sorted[valIndex[i]]];
            s1.pb(i);
            s2.pb(j);
            ans1.pb(val[i]);
            ans2.pb(val[j]);
            swap(val[i], val[j]);
            toIndex[val[j]] = j;
            toIndex[val[i]] = i;
        }
    }

    for (int i = s1.size() - 1; i >= 0; i--) {
        swap(val[s1[i]], val[s2[i]]);
        toIndex[val[s1[i]]] = s1[i];
        toIndex[val[s2[i]]] = s2[i];
    }
    for (int i = 0; i < lo; i++)
        swap(valIndex[X[i]], valIndex[Y[i]]);

    for (int i = 0; i < lo; i++) {
        swap(val[X[i]], val[Y[i]]);
        toIndex[val[X[i]]] = X[i];
        toIndex[val[Y[i]]] = Y[i];

        if (i >= (int)ans1.size())
            continue;

        int si = toIndex[ans1[i]];
        int sj = toIndex[ans2[i]];

        ans1[i] = si;
        ans2[i] = sj;

        swap(val[si], val[sj]);
        toIndex[val[si]] = si;
        toIndex[val[sj]] = sj;
    }

    for (int i = 0; i < M; i++)
        P[i] = Q[i] = 0;

    for (int i = 0; i < min((int)s1.size(), lo); i++)
        P[i] = ans1[i], Q[i] = ans2[i];

    return lo;
}

int main () {
	// freopen("in.txt", "r", stdin);
	// freopen("out.txt", "w", stdout);

    int N, M;
    scanf("%d", &N);

    int S[N];

    for (int i = 0; i < N; i++) {
        scanf("%d", &S[i]);
    }

    scanf("%d", &M);

    int X[M], Y[M];

    for (int i = 0; i < M; i++)
        scanf("%d%d", &X[i], &Y[i]);

    int P[M], Q[M];

    int ans = findSwapPairs(N, S, M, X, Y, P, Q);

    printf("%d\n", ans);
    for (int i = 0; i < ans; i++)
        printf("%d %d\n", P[i], Q[i]);
	return 0;
}

File: src/contest/ioi/IOI_2015_Sorting.java
package contest.ioi;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.StringTokenizer;

public class IOI_2015_Sorting {

  static BufferedReader br;
  static PrintWriter out;
  static StringTokenizer st;

  static int n, m;
  static int[] val, index, x, y;
  static int[] sorted;

  static HashMap<Integer, Integer> toIndex = new HashMap<Integer, Integer>(); // val to index

  public static void main (String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    out = new PrintWriter(new OutputStreamWriter(System.out));
    //br = new BufferedReader(new FileReader("in.txt"));
    //out = new PrintWriter(new FileWriter("out.txt"));

    n = readInt();

    val = new int[n];
    index = new int[n]; // element at index i has "real" index index[i]

    for (int i = 0; i < n; i++) {
      val[i] = readInt();
      index[i] = i;
      toIndex.put(val[i], i);
    }
    sorted = Arrays.copyOf(val, n);
    Arrays.sort(sorted);
    m = readInt();
    x = new int[m];
    y = new int[m];
    for (int i = 0; i < m; i++) {
      x[i] = readInt();
      y[i] = readInt();
    }
    int lo = 0, hi = n - 1;
    ArrayList<Integer> s1 = new ArrayList<Integer>();
    ArrayList<Integer> s2 = new ArrayList<Integer>();
    while (lo <= hi) {
      int mid = (hi + lo) / 2;
      for (int i = mid - 1; i >= 0; i--)
        swap(index, x[i], y[i]);

      int cnt = 0;
      for (int j = 0; j < n; j++) {
        if (sorted[index[j]] != val[j]) {
          int toSwap = toIndex.get(sorted[index[j]]);
          s1.add(j);
          s2.add(toSwap);
          swap(val, j, toSwap);
          toIndex.put(val[j], j);
          toIndex.put(val[toSwap], toSwap);
          cnt++;
        }
      }

      for (int i = s1.size() - 1; i >= 0; i--) {
        swap(val, s1.get(i), s2.get(i));
        toIndex.put(val[s2.get(i)], s2.get(i));
        toIndex.put(val[s1.get(i)], s1.get(i));
      }

      for (int i = 0; i < mid; i++)
        swap(index, x[i], y[i]);

      if (cnt <= mid)
        hi = mid - 1;
      else
        lo = mid + 1;

      s1.clear();
      s2.clear();
    }

    // rebuilding sequence

    ArrayList<Integer> ans1 = new ArrayList<Integer>();
    ArrayList<Integer> ans2 = new ArrayList<Integer>();
    for (int i = lo - 1; i >= 0; i--)
      swap(index, x[i], y[i]);

    for (int j = 0; j < n; j++) {
      if (sorted[index[j]] != val[j]) {
        int toSwap = toIndex.get(sorted[index[j]]);
        s1.add(j);
        s2.add(toSwap);
        ans1.add(val[j]);
        ans2.add(val[toSwap]);
        swap(val, j, toSwap);
        toIndex.put(val[j], j);
        toIndex.put(val[toSwap], toSwap);
      }
    }

    for (int i = s1.size() - 1; i >= 0; i--) {
      swap(val, s1.get(i), s2.get(i));
      toIndex.put(val[s2.get(i)], s2.get(i));
      toIndex.put(val[s1.get(i)], s1.get(i));
    }

    for (int i = 0; i < lo; i++)
      swap(index, x[i], y[i]);

    for (int i = 0; i < lo; i++) {
      swap(val, x[i], y[i]);
      toIndex.put(val[x[i]], x[i]);
      toIndex.put(val[y[i]], y[i]);

      if (i >= ans1.size())
        continue;

      int si = toIndex.get(ans1.get(i));
      int sj = toIndex.get(ans2.get(i));
      ans1.set(i, si);
      ans2.set(i, sj);
      swap(val, si, sj);
      toIndex.put(val[si], si);
      toIndex.put(val[sj], sj);
    }

    out.println(lo);
    for (int i = 0; i < lo; i++) {
      if (i < s1.size())
        out.printf("%d %d\n", ans1.get(i), ans2.get(i));
      else
        out.printf("0 0\n");
    }
    out.close();
  }

  // changing the indexes
  static void swap (int[] a, int i, int j) {
    int temp = a[i];
    a[i] = a[j];
    a[j] = temp;
  }

  static String next () throws IOException {
    while (st == null || !st.hasMoreTokens())
      st = new StringTokenizer(br.readLine().trim());
    return st.nextToken();
  }

  static long readLong () throws IOException {
    return Long.parseLong(next());
  }

  static int readInt () throws IOException {
    return Integer.parseInt(next());
  }

  static double readDouble () throws IOException {
    return Double.parseDouble(next());
  }

  static char readCharacter () throws IOException {
    return next().charAt(0);
  }

  static String readLine () throws IOException {
    return br.readLine().trim();
  }
}
